C++ API parity: MaxUnpool2d

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27060 C++ API parity: LogSigmoid
* #27059 C++ API parity: LeakyReLU
* #27038 C++ API parity: Hardtanh
* #27035 C++ API parity: Hardshrink
* #27028 C++ API parity: ELU
* #27027 C++ API parity: MaxUnpool3d
* **#26915 C++ API parity: MaxUnpool2d**

Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

pbelevich committed:
C++ API parity: MaxUnpool2d

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"


Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"


Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"


Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"


Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"


Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

[ghstack-poisoned]

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -572,6 +572,39 @@ TEST_F(ModulesTest, MaxUnpool1d) {
+      { 0, 41,  0, 43, 44},
+      { 0, 46,  0, 48, 49}}}} , torch::kFloat)));
+  ASSERT_EQ(y.sizes(), torch::IntArrayRef({2, 1, 5, 5}));
+}
```
It would be great to add a MaxPool2d -> MaxUnpool2d test, such as:
```cpp
  {
    MaxPool2d pool(MaxPool2dOptions(2).stride(2));
    MaxUnpool2d unpool(MaxUnpool2dOptions(2).stride(2));
    auto input = torch::tensor({{{{ 1.,  2.,  3.,  4.},
                                  { 5.,  6.,  7.,  8.},
                                  { 9., 10., 11., 12.},
                                  {13., 14., 15., 16.}}}});
    torch::Tensor output, indices;
    std::tie(output, indices) = pool->forward_with_indices(input);
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices),
      torch::tensor({{{{  0.,   0.,   0.,   0.},
                       {  0.,   6.,   0.,   8.},
                       {  0.,   0.,   0.,   0.},
                       {  0.,  14.,   0.,  16.}}}})));
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices, /*output_size=*/{1, 1, 5, 5}),
      torch::tensor({{{{  0.,   0.,   0.,   0.,   0.},
                       {  6.,   0.,   8.,   0.,   0.},
                       {  0.,   0.,   0.,  14.,   0.},
                       { 16.,   0.,   0.,   0.,   0.},
                       {  0.,   0.,   0.,   0.,   0.}}}})));
  }
```

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"


Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

[ghstack-poisoned]

yf225 reviewed (approved):
Thanks @pbelevich !

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"


Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"


Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MaxUnpool2d"


Differential Revision: [D17627826](https://our.internmc.facebook.com/intern/diff/D17627826)

[ghstack-poisoned]
